primitives/benches/pcs.rs

```diff
@@ -4,100 +4,136 @@
-                    nv,
-                    start.elapsed().as_nanos() / repetition as u128
-                );
-
-                MultilinearKzgPCS::commit(&ck, &poly)?
-            };
-
-            // open
-            let (proof, value) = {
-                let start = Instant::now();
-                for _ in 0..repetition {
-                    let _open = MultilinearKzgPCS::open(&ck, &poly, &point)?;
-                }
-
-                println!(
-                    "KZG open for {} variables: {} ns",
-                    nv,
-                    start.elapsed().as_nanos() / repetition as u128
-                );
-                MultilinearKzgPCS::open(&ck, &poly, &point)?
-            };
-
-            // verify
-            {
-                let start = Instant::now();
-                for _ in 0..repetition {
-                    assert!(MultilinearKzgPCS::verify(
-                        &vk, &com, &point, &value, &proof
-                    )?);
-                }
-                println!(
-                    "KZG verify for {} variables: {} ns",
-                    nv,
-                    start.elapsed().as_nanos() / repetition as u128
-                );
-            }
-
-            println!("====================================");
-        }
-
-        Ok(())
-    }
+/// Report the time cost of an opening
+pub fn open<E: Pairing>(
+    pp: &<MultilinearKzgPCS<E> as PolynomialCommitmentScheme>::SRS,
+    num_vars: usize,
+) -> Duration {
+    let rng = &mut test_rng();
+
+    let (ml_ck, _ml_vk) = pp.0.trim(num_vars).unwrap();
+    let (uni_ck, _uni_vk) = pp.1.trim(num_vars).unwrap();
+    let ck = (ml_ck, uni_ck);
+
+    let poly = MLE::from(DenseMultilinearExtension::rand(num_vars, rng));
+    let point: Vec<_> = (0..num_vars).map(|_| E::ScalarField::rand(rng)).collect();
+
+    let start = Instant::now();
+    let _ = MultilinearKzgPCS::open(&ck, &poly, &point).unwrap();
+    start.elapsed()
 }
+
+/// Report the time cost of a verification
+pub fn verify<E: Pairing>(
+    pp: &<MultilinearKzgPCS<E> as PolynomialCommitmentScheme>::SRS,
+    num_vars: usize,
+) -> Duration {
+    let rng = &mut test_rng();
+
+    let (ml_ck, ml_vk) = pp.0.trim(num_vars).unwrap();
+    let (uni_ck, uni_vk) = pp.1.trim(num_vars).unwrap();
+    let ck = (ml_ck, uni_ck);
+    let vk = (ml_vk, uni_vk);
+
+    let poly = MLE::from(DenseMultilinearExtension::rand(num_vars, rng));
+    let point: Vec<_> = (0..num_vars).map(|_| E::ScalarField::rand(rng)).collect();
+
+    let commitment = MultilinearKzgPCS::commit(&ck, &poly).unwrap();
+
+    let (proof, value) = MultilinearKzgPCS::open(&ck, &poly, &point).unwrap();
+
+    let start = Instant::now();
+    assert!(MultilinearKzgPCS::verify(&vk, &commitment, &point, &value, &proof).unwrap());
+    start.elapsed()
+}
+
+fn kzg_254(c: &mut Criterion) {
+    bench_pcs_method::<Bn254>(
+        c,
+        (MIN_NUM_VARS..MAX_NUM_VARS).step_by(2).collect(),
```

Comment:
Instead of collecting it into a Vec, can we change this function to accept [Range](https://doc.rust-lang.org/std/ops/struct.Range.html) ?

Reply:
Good call, thank you. We tried that at some point and it failed because `step_by` returns a `StepBy` object which, oddly enough, cannot immediately be converted into `Range`.

However, `StepBy` does implement `Iterator` and so I have changed the argument to `bench_pcs_method` to expect an implementor of that - which is also more general.